fl4health/model_bases/autoencoders_base.py

```diff
@@ -0,0 +1,162 @@
+
+
+class ConditionalVAE(AbstractAE):
+    def __init__(
+        self,
+        model_type: AutoEncoderType,
+        num_conditions: int,
+        encoder: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        """Initializes the ConditionalVAE class.
+
+        Args:
+            model_type (AutoEncoderType): The model type should be CONDITIONAL_VAE.
+            num_conditions (int): The total number of conditions, which is used to prepare the condition vector.
+            encoder (nn.Module): Encoder module defined by the user.
+            decoder (nn.Module): Decoder module defined by the user.
+        """
+        super().__init__(model_type, encoder, decoder)
+        self.num_conditions = num_conditions
+
+    def cat_input_condition(self, input: torch.Tensor, condition: torch.Tensor) -> torch.Tensor:
+        input_cond = torch.cat((input, condition), dim=-1)
+        return input_cond
+
+    def one_hot(self, condition: torch.Tensor) -> torch.Tensor:
+        # One-hot encoding of the condition
+        condition = F.one_hot(condition.to(torch.int64), self.num_conditions).to(condition.device)
+        return condition
+
+    def maybe_reshape(self, condition_matrix: torch.Tensor) -> torch.Tensor:
+        # Make sure the condition has the right shape.
+        if len(condition_matrix.shape) > 2:
+            # Since the condition is replicated to match the size of the input,
+            #  we need to narrow it back just to one number per sample
+            return condition_matrix[:, 0, 0]
+        else:
+            # Condition was not replicated to match the size of the input.
+            return condition_matrix
+
+    def encode(self, input: torch.Tensor, condition: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        # User can decide how to use the condition in the encoder,
+        # ex: using the condition in the middle layers of encoder.
+        mu, logvar = self.encoder(input, condition)
+        return mu, logvar
+
+    def decode(self, latent_vector: torch.Tensor, condition: torch.Tensor) -> torch.Tensor:
+        latent_vector = self.cat_input_condition(latent_vector, condition)
+        output = self.decoder(latent_vector)
+        return output
+
+    def sampling(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
+        std = torch.exp(0.5 * logvar)
+        eps = torch.randn_like(std)
+        return eps.mul(std).add_(mu)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        x = input[:, :-1]  # Exclude the last item (conditions)
+        condition_matrix = input[:, -1]  # Last item contains conditions
+        condition = self.maybe_reshape(condition_matrix)
+        one_hot_condition = self.one_hot(condition)
+        mu, logvar = self.encode(x, one_hot_condition)
+        z = self.sampling(mu, logvar)
+        output = self.decode(z, one_hot_condition)
+        # The shape of the flattened_output can be later restored by having the training image shape,
```

Comment:
This comment is "image" data specific. I'd suggest rephrasing to be a little more general.